src/Symfony/Component/String/Uuid.php

```diff
@@ -0,0 +1,133 @@
+
+/**
+ * @author Grégoire Pineau <lyrixx@lyrixx.info>
+ */
+final class Uuid implements \JsonSerializable
+{
+    public const TYPE_1 = UUID_TYPE_TIME;
+    public const TYPE_3 = UUID_TYPE_MD5;
+    public const TYPE_4 = UUID_TYPE_RANDOM;
+    public const TYPE_5 = UUID_TYPE_SHA1;
+
+    public const VARIANT_NCS = UUID_VARIANT_NCS;
+    public const VARIANT_DCE = UUID_VARIANT_DCE;
+    public const VARIANT_MICROSOFT = UUID_VARIANT_MICROSOFT;
+    public const VARIANT_OTHER = UUID_VARIANT_OTHER;
+
+    private $uuid;
+
+    public function __construct(string $uuid)
+    {
+        self::checkIfExtensionIsLoaded();
+
+        if (!uuid_is_valid($uuid)) {
+            throw new \InvalidArgumentException('The UUID is not valid.');
+        }
+
+        $this->uuid = $uuid;
+    }
+
+    public static function v1(): self
+    {
+        self::checkIfExtensionIsLoaded();
+
+        return new self(uuid_create(UUID_TYPE_TIME));
+    }
+
+    public static function v3(self $uuidNamespace, string $name): self
+    {
+        self::checkIfExtensionIsLoaded();
+
+        return new self(uuid_generate_md5($uuidNamespace->uuid, $name));
+    }
+
+    public static function v4(): self
+    {
+        self::checkIfExtensionIsLoaded();
+
+        return new self(uuid_create(UUID_TYPE_RANDOM));
+    }
+
+    public static function v5(self $uuidNamespace, string $name): self
+    {
+        self::checkIfExtensionIsLoaded();
+
+        return new self(uuid_generate_sha1($uuidNamespace->uuid, $name));
+    }
+
+    public static function fromBinary(string $uuidAsBinary): self
+    {
+        self::checkIfExtensionIsLoaded();
+
+        return new self(uuid_unparse($uuidAsBinary));
+    }
+
+    public function toBinary(): string
+    {
+        return uuid_parse($this->uuid);
+    }
+
+    public function isNull(): bool
+    {
+        return uuid_is_null($this->uuid);
+    }
+
+    public function isEqualsTo(self $other): bool
+    {
+        return $this->uuid === $other->uuid;
+    }
+
+    public function compare(self $other): int
+    {
+        return uuid_compare($this->uuid, $other->uuid);
+    }
+
+    public function getType(): int
+    {
+        return uuid_type($this->uuid);
+    }
+
+    public function getVariant(): int
+    {
+        return uuid_variant($this->uuid);
+    }
+
+    public function getTime(): int
+    {
+        return uuid_time($this->uuid);
+    }
+
+    public function getMac(): string
+    {
+        return uuid_mac($this->uuid);
+    }
+
+    public function __toString()
+    {
+        return $this->uuid;
+    }
+
+    public function jsonSerialize(): string
+    {
+        return $this->uuid;
+    }
+
+    private static function checkIfExtensionIsLoaded()
+    {
+        if (!\function_exists('uuid_is_valid')) {
+            throw new \LogicException(sprintf('You cannot use the "%s" class as the UUID extension is not installed. Try running "composer require symfony/polyfill-uuid".', __CLASS__));
```

Comment:
Not sure if this is the best practice of the Symfony components, but should t this LogicException be part of the String component instead of the global one?

Reply:
`LogicException` are like `Throwable`: they should not be caught and should bubble to up to the error page instead. Which means no need for a dedicated exception class (as done already in many components.)

Reply:
Thanks for the explanation! Same for \InvalidArgumentException ?

Reply:
`InvalidArgumentException extends LogicException` ;)